### PR TITLE
Omit adding the `osd-version` header when the Fetch request is to an external origin

### LINKED ISSUE #3277: [BUG] Fetch calls pass custom headers for every 302 request causing CORS
**Describe the bug**

In some scenarios where a 302 redirect request is received by OpenSearch Dashboards, it makes a fetch call to the redirected url using custom headers like `osd-version` [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/http/fetch.ts#L147). If the redirected URL doesn't belong to OpenSearch Dashboards and is meant for an external authentication system, such custom headers can cause the browser to treat the request as eligible for [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests). 

This causes the browser to first send a preflight request to the redirected endpoint to check whether custom headers are allowed.  An example is `Access-Control-Request-Headers: osd-version`

**To Reproduce**
Not providing the exact steps to reproduce as this involves the use of external authentication systems and associated plugins present in OS core written to intercept and authenticate requests which are proprietary. 

- User tries to access dashboards with an expired session. 
- Request to OpenSearch Dashboards `GET capabilities API` is intercepted by proprietary plugin and it sees that token in cookie is expired.
- Proprietary plugin redirects the user to external authentication system.
- Redirected URL is loaded by OSD using `fetch.ts` and custom header of `osd-version` is passed. 
- Browser makes a Pre-flight `OPTIONS` request to external authentication system due to the presence of custom header.
- Pre-flight request to the external authentication system fails and user is not redirected to login page of external authentication system.  

**Expected behavior**

Since, the custom header `osd-version` is only understood by OSD, we should avoid sending it if the redirected URL doesn't belong to Dashboards.  

**OpenSearch Version**
1.3

**Dashboards Version**
1.3

**Plugins**
- alertingDashboards
- anomalyDetectionDashboards
- cognitoLogout
- ganttChartDashboards
- indexManagementDashboards
- observabilityDashboards
- queryWorkbenchDashboards
- reportsDashboards
- securityDashboards
- storage

**Host/Environment (please complete the following information):**
 - iOS
 - Chrome version 106

**Additional context**

OpenSearch Dashboards supports multiple ways of authentication for a user like Basic Auth, SAML, OpenID; etc. Most of these are natively supported by the security-dashboards-plugin and security-plugin. 

To enable other forms of external authentication methods that are not currently supported in open-source, installations can make use of plugins in Opensearch to intercept and authenticate requests. Such external authentication systems can perform redirection to external login endpoint if the credentials of the user is expired. 

Such redirections currently fail to succeed, due to the presence of custom headers like `osd-version` in the redirected request added by OSD core.

### PR DESCRIPTION
**Pull request by AMoo-Miki**
### Description

* Making `fetch` requests using core/public/http/fetch, an `osd-version` header is forcefully added, even to external requests. This change examines the destination and only adds the header to relative URLs and those that are to OSD itself.
* This change also adds `osd-xsrf` to calls that use `osd-version` incorrectly to satisfy XSRF protection
 
### Issues Resolved
Fixes #3277
 
### Check List
- [X] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [X] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by kavilla — CHANGES_REQUESTED** [review]
I think the failures since we utilized this for XRSF.

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/de0634495ebace79453b50421e8d147104a7b562/src/core/server/http/lifecycle_handlers.ts#L57 tries to find a header and then it can't find it so it throws the error.

It might not be the proper way of handling XSRF but by removing it causes these errors to be thrown.

**Comment by AMoo-Miki** [comment]
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/de0634495ebace79453b50421e8d147104a7b562/src/core/server/http/lifecycle_handlers.ts#L57

If it is a `GET` or `OPTIONS` call, in will not fail and also if it contains the `osd-xsrf`. https://github.com/opensearch-project/OpenSearch-Dashboards/commit/621bf0efec3e96884b86d94462b1698171933f87 suggests `osd-version` used to be used for XSRF prior to K5 and they never cared to clean the code up over the next 6 years. I have added `osd-xsrf` to any call that was using `osd-version` for XSRF protection.

**Comment by joshuarrrr** [comment]
@AMoo-Miki There are some snapshots that need updating.

**Review by kavilla — APPROVED** [review]
I think looks good to me. Pulled it down working. It's also working with the basic settings for security plugin.

One thing I'm not able to completely test is if any functionality relied on the `osd-version` header being present just so the redirect can be attached to the request and not fail.

For example, on `main`:
This succeeds:
```
curl localhost:5601/hdn/api/status -ku 'admin:admin'
```
This succeeds:
```
curl localhost:5601/hdn/api/status -ku 'admin:admin' -H osd-version:3.0.0
```
This succeeds:
```
curl localhost:5601/hdn/api/status -ku 'admin:admin' -H osd-version:
```
This fails:
```
curl localhost:5601/hdn/api/status -ku 'admin:admin' -H osd-version:3.5.0
```
Which makes sense because the version mismatch. So if any functionality outside of what I tested relied on the osd-version and utilized the fetch call might be broken if the fetch was passed from one request to another. 

So I don't know want to block it on something that I can't test especially if the value being not present still works for us. But I do believe we might want to mention in higher in the changelog.

Like something that mentions,

HTTP Fetch from Core OpenSearch Dashboards no longer default appends the `osd-version`.

**Comment by AMoo-Miki** [comment]
Rocky, this change makes sure `osd-xsrf` is used all over and only removed `node-version` when the call is to a destination outside of OSD; any call to OSD will have `node-version`.

File: src/core/server/http/lifecycle_handlers.ts [hunk 1]
```diff
@@ -54,8 +54,9 @@ export const createXsrfPostAuthHandler = (config: HttpConfig): OnPostAuthHandler
     const hasVersionHeader = VERSION_HEADER in request.headers;
     const hasXsrfHeader = XSRF_HEADER in request.headers;
 
+    // ToDo: Remove !hasVersionHeader; `osd-version` incorrectly used for satisfying XSRF protection
```
Comment (joshuarrrr) [inline src/core/server/http/lifecycle_handlers.ts:57]: nit - todo comments should have issues opened to link to.

**Review by joshuarrrr — APPROVED** [review]
@AMoo-Miki Is there a follow-up issue already created to keep track of these todos?

**Pull request merged**